File: nuit-core/src/utils/style/shadow.rs
use serde::{Deserialize, Serialize};

use crate::{Color, Vec2};

/// A style for rendered shadows.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Shadow {
    kind: ShadowKind,
    color: Color,
    radius: f64,
    offset: Vec2<f64>,
}

/// A type of shadow.
#[derive(Debug, Default, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum ShadowKind {
    #[default]
    Drop,
    Inner,
}

impl Shadow {
    #[must_use]
    pub const fn new(kind: ShadowKind, color: Color, radius: f64, offset: Vec2<f64>) -> Self {
        Self { kind, color, radius, offset }
    }

    #[must_use]
    pub const fn drop(color: Color, radius: f64, offset: Vec2<f64>) -> Self {
        Self::new(ShadowKind::Drop, color, radius, offset)
    }

    #[must_use]
    pub const fn inner(color: Color, radius: f64, offset: Vec2<f64>) -> Self {
        Self::new(ShadowKind::Inner, color, radius, offset)
    }
}

File: nuit-bridge-swiftui/Sources/NuitBridgeSwiftUICore/Utils/Style/Shadow.swift
public struct Shadow: Codable, Hashable {
    public let kind: Kind
    public let color: Color
    public let radius: Double
    public let offset: Vec2<Double>

    public enum Kind: String, Codable, Hashable {
        case drop
        case inner
    }
    
    public init(kind: Kind, color: Color, radius: Double, offset: Vec2<Double>) {
        self.kind = kind
        self.color = color
        self.radius = radius
        self.offset = offset
    }
}
